diff --git a/node_modules/cesium/Source/Scene/Material.js b/node_modules/cesium/Source/Scene/Material.js
index 59d00a8..ddd8bc7 100644
--- a/node_modules/cesium/Source/Scene/Material.js
+++ b/node_modules/cesium/Source/Scene/Material.js
@@ -762,6 +762,36 @@ import when from '../ThirdParty/when.js';
                 return;
             }
 
+            if (uniformValue instanceof HTMLCanvasElement) {
+
+                if (!uniformValue.drawing) {
+                    if (uniformChanged && defined(texture)) {
+                        if (texture !== context.defaultTexture) {
+                            texture.destroy();
+                        }
+                        texture = undefined;
+                    }
+
+                    if (!defined(texture) || texture === context.defaultTexture) {
+                        texture = new Texture({
+                            context : context,
+                            source : uniformValue
+                        });
+    
+                        material._textures[uniformId] = texture;
+    
+                        return;
+                    }
+    
+                    texture.copyFrom(uniformValue);
+
+                } else if (!defined(texture)) {
+                    material._textures[uniformId] = context.defaultTexture;
+                }
+
+                return;
+            }
+
             if (uniformValue instanceof Texture && uniformValue !== texture) {
                 material._texturePaths[uniformId] = undefined;
                 var tmp = material._textures[uniformId];
