// @loaders.gl/images/dist/es5/image-loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _parseImage = _interopRequireDefault(require("./lib/parsers/parse-image"));

var _binaryImageApi = require("./lib/category-api/binary-image-api");

var VERSION = typeof "2.3.8" !== 'undefined' ? "2.3.8" : 'latest';
var EXTENSIONS = ['png', 'jpg', 'jpeg', 'gif', 'webp', 'bmp', 'ico', 'svg'];
var MIME_TYPES = ['image/png', 'image/jpeg', 'image/gif', 'image/webp', 'image/bmp', 'image/vnd.microsoft.icon', 'image/svg+xml'];
var ImageLoader = {
  id: 'image',
  name: 'Images',
  version: VERSION,
  mimeTypes: MIME_TYPES,
  extensions: EXTENSIONS,
  parse: _parseImage["default"],
  tests: [function (arrayBuffer) {
    return Boolean((0, _binaryImageApi.getBinaryImageMetadata)(new DataView(arrayBuffer)));
  }],
  options: {
    image: {
      type: 'auto',
      decode: true
    }
  }
};
var _default = ImageLoader;
exports["default"] = _default;
