// @deck.gl/geo-layers/dist/es6/mvt-layer/mvt-layer.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { Matrix4 } from 'math.gl';
import { MVTLoader } from '@loaders.gl/mvt';
import { load } from '@loaders.gl/core';
import { COORDINATE_SYSTEM } from '@deck.gl/core';
import TileLayer from '../tile-layer/tile-layer';
import { getURLFromTemplate } from '../tile-layer/utils';
import ClipExtension from './clip-extension';
const WORLD_SIZE = 512;
const defaultProps = {
  uniqueIdProperty: {
    type: 'string',
    value: ''
  },
  highlightedFeatureId: null
};
export default class MVTLayer extends TileLayer {
  getTileData(tile) {
    const url = getURLFromTemplate(this.props.data, tile);

    if (!url) {
      return Promise.reject('Invalid URL');
    }

    let options = this.getLoadOptions();
    options = _objectSpread(_objectSpread({}, options), {}, {
      mvt: _objectSpread(_objectSpread({}, options && options.mvt), {}, {
        coordinates: this.context.viewport.resolution ? 'wgs84' : 'local',
        tileIndex: {
          x: tile.x,
          y: tile.y,
          z: tile.z
        }
      })
    });
    return load(url, MVTLoader, options);
  }

  renderSubLayers(props) {
    const {
      tile
    } = props;
    const worldScale = Math.pow(2, tile.z);
    const xScale = WORLD_SIZE / worldScale;
    const yScale = -xScale;
    const xOffset = WORLD_SIZE * tile.x / worldScale;
    const yOffset = WORLD_SIZE * (1 - tile.y / worldScale);
    const modelMatrix = new Matrix4().scale([xScale, yScale, 1]);
    props.autoHighlight = false;

    if (!this.context.viewport.resolution) {
      props.modelMatrix = modelMatrix;
      props.coordinateOrigin = [xOffset, yOffset, 0];
      props.coordinateSystem = COORDINATE_SYSTEM.CARTESIAN;
      props.extensions = [...(props.extensions || []), new ClipExtension()];
    }

    return super.renderSubLayers(props);
  }

  onHover(info, pickingEvent) {
    const {
      uniqueIdProperty,
      autoHighlight
    } = this.props;

    if (autoHighlight) {
      const {
        hoveredFeatureId
      } = this.state;
      const hoveredFeature = info.object;
      let newHoveredFeatureId;

      if (hoveredFeature) {
        newHoveredFeatureId = getFeatureUniqueId(hoveredFeature, uniqueIdProperty);
      }

      if (hoveredFeatureId !== newHoveredFeatureId && newHoveredFeatureId !== -1) {
        this.setState({
          hoveredFeatureId: newHoveredFeatureId
        });
      }
    }

    return super.onHover(info, pickingEvent);
  }

  getHighlightedObjectIndex(tile) {
    const {
      hoveredFeatureId
    } = this.state;
    const {
      uniqueIdProperty,
      highlightedFeatureId
    } = this.props;
    const {
      data
    } = tile;
    const isFeatureIdPresent = isFeatureIdDefined(hoveredFeatureId) || isFeatureIdDefined(highlightedFeatureId);

    if (!isFeatureIdPresent || !Array.isArray(data)) {
      return -1;
    }

    const featureIdToHighlight = isFeatureIdDefined(highlightedFeatureId) ? highlightedFeatureId : hoveredFeatureId;
    return data.findIndex(feature => getFeatureUniqueId(feature, uniqueIdProperty) === featureIdToHighlight);
  }

}

function getFeatureUniqueId(feature, uniqueIdProperty) {
  if (uniqueIdProperty) {
    return feature.properties[uniqueIdProperty];
  }

  if ('id' in feature) {
    return feature.id;
  }

  return -1;
}

function isFeatureIdDefined(value) {
  return value !== undefined && value !== null && value !== '';
}

MVTLayer.layerName = 'MVTLayer';
MVTLayer.defaultProps = defaultProps;
