diff --git a/node_modules/cesium/Source/Scene/Material.js b/node_modules/cesium/Source/Scene/Material.js	2021-10-02 19:27:27.390680381 +0200
--- a/node_modules/cesium/Source/Scene/Material.js
+++ b/node_modules/cesium/Source/Scene/Material.js
@@ -846,6 +846,33 @@
       return;
     }

+    if (uniformValue instanceof HTMLCanvasElement) {
+      if (!uniformValue.drawing) {
+        if (uniformChanged && defined(texture)) {
+          if (texture !== context.defaultTexture) {
+            texture.destroy();
+          }
+          texture = undefined;
+        }
+
+        if (!defined(texture) || texture === context.defaultTexture) {
+          texture = new Texture({
+            context: context,
+            source: uniformValue,
+          });
+          material._textures[uniformId] = texture;
+          return;
+        }
+
+        texture.copyFrom({
+          source: uniformValue,
+        });
+      } else if (!defined(texture)) {
+        material._textures[uniformId] = context.defaultTexture;
+      }
+      return;
+    }
+
     if (uniformValue instanceof Texture && uniformValue !== texture) {
       material._texturePaths[uniformId] = undefined;
       var tmp = material._textures[uniformId];