// @loaders.gl/gltf/dist/es5/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "KHR_DRACO_MESH_COMPRESSION", {
  enumerable: true,
  get: function get() {
    return _gltfConstants.KHR_DRACO_MESH_COMPRESSION;
  }
});
Object.defineProperty(exports, "UBER_POINT_CLOUD_EXTENSION", {
  enumerable: true,
  get: function get() {
    return _gltfConstants.UBER_POINT_CLOUD_EXTENSION;
  }
});
Object.defineProperty(exports, "GLTFLoader", {
  enumerable: true,
  get: function get() {
    return _gltfLoader["default"];
  }
});
Object.defineProperty(exports, "GLTFWriter", {
  enumerable: true,
  get: function get() {
    return _gltfWriter["default"];
  }
});
Object.defineProperty(exports, "GLBLoader", {
  enumerable: true,
  get: function get() {
    return _glbLoader["default"];
  }
});
Object.defineProperty(exports, "GLBWriter", {
  enumerable: true,
  get: function get() {
    return _glbWriter["default"];
  }
});
Object.defineProperty(exports, "GLTFScenegraph", {
  enumerable: true,
  get: function get() {
    return _gltfScenegraph["default"];
  }
});
Object.defineProperty(exports, "postProcessGLTF", {
  enumerable: true,
  get: function get() {
    return _postProcessGltf["default"];
  }
});
Object.defineProperty(exports, "GLBBuilder", {
  enumerable: true,
  get: function get() {
    return _glbBuilder["default"];
  }
});
Object.defineProperty(exports, "encodeGLTFSync", {
  enumerable: true,
  get: function get() {
    return _encodeGltf.encodeGLTFSync;
  }
});

var _gltfConstants = require("./lib/gltf-constants");

var _gltfLoader = _interopRequireDefault(require("./gltf-loader"));

var _gltfWriter = _interopRequireDefault(require("./gltf-writer"));

var _glbLoader = _interopRequireDefault(require("./glb-loader"));

var _glbWriter = _interopRequireDefault(require("./glb-writer"));

var _gltfScenegraph = _interopRequireDefault(require("./lib/gltf-scenegraph"));

var _postProcessGltf = _interopRequireDefault(require("./lib/post-process-gltf"));

var _glbBuilder = _interopRequireDefault(require("./lib/deprecated/glb-builder"));

var _encodeGltf = require("./lib/encode-gltf");
