diff --git a/node_modules/nouislider/distribute/nouislider.js b/node_modules/nouislider/distribute/nouislider.js
--- a/node_modules/nouislider/distribute/nouislider.js
+++ b/node_modules/nouislider/distribute/nouislider.js
@@ -1465,9 +1465,9 @@
                 }

                 // Stop if an active 'tap' transition is taking place.
-                if (hasClass(scope_Target, options.cssClasses.tap) && !data.doNotReject) {
-                    return false;
-                }
+                // if (hasClass(scope_Target, options.cssClasses.tap) && !data.doNotReject) {
+                //     return false;
+                // }

                 // Ignore right or middle clicks on start #454
                 if (events === actions.start && e.buttons !== undefined && e.buttons > 1) {
