// @luma.gl/gltools/dist/es5/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "assert", {
  enumerable: true,
  get: function get() {
    return _utils.assert;
  }
});
Object.defineProperty(exports, "deepArrayEqual", {
  enumerable: true,
  get: function get() {
    return _utils.deepArrayEqual;
  }
});
Object.defineProperty(exports, "isObjectEmpty", {
  enumerable: true,
  get: function get() {
    return _utils.isObjectEmpty;
  }
});
Object.defineProperty(exports, "cssToDevicePixels", {
  enumerable: true,
  get: function get() {
    return _devicePixels.cssToDevicePixels;
  }
});
Object.defineProperty(exports, "cssToDeviceRatio", {
  enumerable: true,
  get: function get() {
    return _devicePixels.cssToDeviceRatio;
  }
});
Object.defineProperty(exports, "getDevicePixelRatio", {
  enumerable: true,
  get: function get() {
    return _devicePixels.getDevicePixelRatio;
  }
});
Object.defineProperty(exports, "isWebGL", {
  enumerable: true,
  get: function get() {
    return _webglChecks.isWebGL;
  }
});
Object.defineProperty(exports, "isWebGL2", {
  enumerable: true,
  get: function get() {
    return _webglChecks.isWebGL2;
  }
});
exports.log = void 0;

var _probe = require("probe.gl");

var _utils = require("./utils");

var _devicePixels = require("./device-pixels");

var _webglChecks = require("./webgl-checks");

var log = new _probe.Log({
  id: 'luma.gl'
});
exports.log = log;
