// @loaders.gl/gltf/dist/es5/lib/gltf-constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getBytesFromComponentType = getBytesFromComponentType;
exports.getSizeFromAccessorType = getSizeFromAccessorType;
exports.getGLEnumFromSamplerParameter = getGLEnumFromSamplerParameter;
exports.UBER_POINT_CLOUD_EXTENSION = exports.KHR_TECHNIQUES_WEBGL = exports.KHR_MATERIALS_UNLIT = exports.KHR_LIGHTS_PUNCTUAL = exports.KHR_DRACO_MESH_COMPRESSION = exports.KHR_BINARY_GLTF = void 0;
var KHR_BINARY_GLTF = 'KHR_binary_glTF';
exports.KHR_BINARY_GLTF = KHR_BINARY_GLTF;
var KHR_DRACO_MESH_COMPRESSION = 'KHR_draco_mesh_compression';
exports.KHR_DRACO_MESH_COMPRESSION = KHR_DRACO_MESH_COMPRESSION;
var KHR_LIGHTS_PUNCTUAL = 'KHR_lights_punctual';
exports.KHR_LIGHTS_PUNCTUAL = KHR_LIGHTS_PUNCTUAL;
var KHR_MATERIALS_UNLIT = 'KHR_materials_unlit';
exports.KHR_MATERIALS_UNLIT = KHR_MATERIALS_UNLIT;
var KHR_TECHNIQUES_WEBGL = 'KHR_techniques_webgl';
exports.KHR_TECHNIQUES_WEBGL = KHR_TECHNIQUES_WEBGL;
var UBER_POINT_CLOUD_EXTENSION = 'UBER_draco_point_cloud_compression';
exports.UBER_POINT_CLOUD_EXTENSION = UBER_POINT_CLOUD_EXTENSION;
var COMPONENTS = {
  SCALAR: 1,
  VEC2: 2,
  VEC3: 3,
  VEC4: 4,
  MAT2: 4,
  MAT3: 9,
  MAT4: 16
};
var BYTES = {
  5120: 1,
  5121: 1,
  5122: 2,
  5123: 2,
  5125: 4,
  5126: 4
};

function getBytesFromComponentType(componentType) {
  return BYTES[componentType];
}

function getSizeFromAccessorType(type) {
  return COMPONENTS[type];
}

function getGLEnumFromSamplerParameter(parameter) {
  var GL_TEXTURE_MAG_FILTER = 0x2800;
  var GL_TEXTURE_MIN_FILTER = 0x2801;
  var GL_TEXTURE_WRAP_S = 0x2802;
  var GL_TEXTURE_WRAP_T = 0x2803;
  var PARAMETER_MAP = {
    magFilter: GL_TEXTURE_MAG_FILTER,
    minFilter: GL_TEXTURE_MIN_FILTER,
    wrapS: GL_TEXTURE_WRAP_S,
    wrapT: GL_TEXTURE_WRAP_T
  };
  return PARAMETER_MAP[parameter];
}
